backend/apid/actions/users.go

```diff
@@ -74,17 +78,32 @@ func (a UserController) CreateOrReplace(ctx context.Context, user *corev2.User)
 		return NewError(InvalidArgument, err)
 	}
 
-	// Validate password
-	if err := user.ValidatePassword(); err != nil {
-		return NewError(InvalidArgument, err)
-	}
+	// Determine if a hashed and/or cleartext password was provided
+	if user.Password != "" && user.PasswordHash != "" {
+		// Both the cleartext & hashed passwords were provided, so we need to make
+		if ok := bcrypt.CheckPassword(user.PasswordHash, user.Password); !ok {
+			return NewError(InvalidArgument, errors.New("hashed password does not the match the password"))
+		}
+		user.Password = ""
+	} else if user.Password != "" {
+		// We need to validate the cleartext passsword so it matches our minimal
+		// requirements
+		if err := user.ValidatePassword(); err != nil {
+			return NewError(InvalidArgument, err)
+		}
 
-	// Create password digest
-	hash, err := bcrypt.HashPassword(user.Password)
-	if err != nil {
-		return NewError(InternalErr, err)
+		// Create a hash for this password
+		hash, err := bcrypt.HashPassword(user.Password)
+		if err != nil {
+			return NewError(InternalErr, err)
+		}
+		user.PasswordHash = hash
+
+		// Override the cleartext password so we don't store it
+		user.Password = ""
```

Comment:
I think we should not do this. Instead, we should store the hashed password in `user.Password`.

Reply:
So you're saying we should the hashed password in both `user.Password` & `user.PasswordHash`?

Reply:
Yes, otherwise users who get created after this change, or change their password, can never be authenticated if the version is rolled back.

Reply:
I see, it makes sense.